src/perf/reader.cpp

```diff
@@ -0,0 +1,344 @@
+{
+    if (config().use_pebs)
+    {
+        attr_.use_clockid = 0;
+    }
+
+    attr_.sample_period = config().sampling_period;
+
+    if (config().sampling)
+    {
+        Log::debug() << "using sampling event \'" << config().sampling_event
+                     << "\', period: " << config().sampling_period;
+
+        attr_.mmap = 1;
+    }
+    else
+    {
+        // Set up a dummy event for recording calling context enter/leaves only
+        attr_.type = PERF_TYPE_SOFTWARE;
+        attr_.config = PERF_COUNT_SW_DUMMY;
+    }
+
+    attr_.sample_id_all = 1;
+    // Generate PERF_RECORD_COMM events to trace changes to the command
+    // name of a task.  This is used to write a meaningful name for any
+    // traced thread to the archive.
+    attr_.comm = 1;
+    attr_.context_switch = 1;
+
+    // TODO see if we can remove remove tid
+    attr_.sample_type |= PERF_SAMPLE_IP | PERF_SAMPLE_TID | PERF_SAMPLE_CPU;
+    if (config().enable_cct)
+    {
+        attr_.sample_type |= PERF_SAMPLE_CALLCHAIN;
+    }
+
+    attr_.precise_ip = 3;
+}
+
+const std::set<Cpu>& SysfsEvent::supported_cpus() const
+{
+    if (cpus_.empty())
+    {
+        return Topology::instance().cpus();
+    }
+    return cpus_;
+}
+
+bool SysfsEvent::degrade_percision()
+{
+    /* reduce exactness of IP can help if the kernel does not support really exact events */
+    if (attr_.precise_ip == 0)
+    {
+        return false;
+    }
+    else
+    {
+        attr_.precise_ip--;
+        return true;
+    }
+}
+
+bool SysfsEvent::is_valid() const
+{
+    return (availability_ != Availability::UNAVAILABLE);
+}
+
+void SysfsEvent::set_availability()
+{
+    PerfEventInstance proc_ev;
+    PerfEventInstance sys_ev;
+
+    try
+    {
+        proc_ev = open(Thread(0), -1, -1);
+    }
+    catch (const std::system_error& e)
+    {
+    }
+
+    try
+    {
+        sys_ev = open(*supported_cpus().begin(), -1, -1);
+    }
+    catch (const std::system_error& e)
+    {
+    }
+
+    if (proc_ev.get_fd() == -1)
+    {
+        if (sys_ev.get_fd() == -1)
+            availability_ = Availability::UNAVAILABLE;
+        else
+            availability_ = Availability::SYSTEM_MODE;
+    }
+    else
+    {
+        if (sys_ev.get_fd() == -1)
+            availability_ = Availability::PROCESS_MODE;
+        else
+            availability_ = Availability::UNIVERSAL;
+    }
+}
+
+TimeEvent::TimeEvent(bool enable_on_exec, [[maybe_unused]] uint64_t addr)
+: PerfEvent(enable_on_exec)
+{
+    attr_.exclude_kernel = 1;
+
+#ifndef USE_HW_BREAKPOINT_COMPAT
+    attr_.type = PERF_TYPE_BREAKPOINT;
+    attr_.bp_type = HW_BREAKPOINT_W;
+    attr_.bp_len = HW_BREAKPOINT_LEN_8;
+    attr_.wakeup_events = 1;
+    attr_.bp_addr = addr;
+#else
+    attr_.type = PERF_TYPE_HARDWARE;
+    attr_.config = PERF_COUNT_HW_INSTRUCTIONS;
+    attr_.sample_period = 100000000;
+    attr_.task = 1;
+#endif
+}
+
+TracepointEvent::TracepointEvent(bool enable_on_exec, int event_id) : PerfEvent(enable_on_exec)
```

Comment:
I  _feel_ there should be a TracepointEvent constructor that just takes the name of the tracepoint and does the tracepoint parsing stuff internally.

Reply:
Should TracepointEvent then replace the whole EventFormat class?

Reply:
ideally, yes